icelake/src/io/file_writer/equality_delta_writer.rs

```diff
@@ -0,0 +1,247 @@
+    row_converter: RowConverter,
+    primary_column_idx: Vec<usize>,
+}
+
+impl EqualityDeltaWriter {
+    /// Create a new `EqualityDeltaWriter`.
+    pub async fn try_new(
+        operator: Operator,
+        table_location: String,
+        data_location_generator: Arc<FileLocationGenerator>,
+        delete_location_generator: Arc<FileLocationGenerator>,
+        arrow_schema: SchemaRef,
+        table_config: TableConfigRef,
+        primary_column_ids: Vec<usize>,
+    ) -> Result<Self> {
+        let mut primary_column_idx = vec![];
+        for &id in primary_column_ids.iter() {
+            arrow_schema.fields().iter().enumerate().any(|(idx, f)| {
+                if f.metadata()
+                    .get(COLUMN_ID_META_KEY)
+                    .unwrap()
+                    .parse::<usize>()
+                    .unwrap()
+                    == id
+                {
+                    primary_column_idx.push(idx);
+                    true
+                } else {
+                    false
+                }
+            });
+        }
+        let primary_schema = arrow_schema.project(&primary_column_idx).map_err(|err| {
+            Error::new(
+                ErrorKind::ArrowError,
+                format!(
+                    "Failed to project schema with equality ids: {:?}, error: {}",
+                    primary_column_idx, err
+                ),
+            )
+        })?;
+        let row_converter = RowConverter::new(
+            primary_schema
+                .fields()
+                .iter()
+                .map(|f| SortField::new(f.data_type().clone()))
+                .collect(),
+        )
+        .map_err(|err| {
+            Error::new(
+                ErrorKind::ArrowError,
+                format!("Failed to create row converter, error: {}", err),
+            )
+        })?;
+        Ok(Self {
+            data_file_writer: DataFileWriter::try_new(
+                operator.clone(),
+                table_location.clone(),
+                data_location_generator.clone(),
+                arrow_schema.clone(),
+                table_config.clone(),
+            )
+            .await?,
+            sorted_pos_delete_writer: SortedPositionDeleteWriter::new(
+                operator.clone(),
+                table_location.clone(),
+                delete_location_generator.clone(),
+                table_config.clone(),
+            ),
+            eq_delete_writer: EqualityDeleteWriter::try_new(
+                operator,
+                table_location,
+                delete_location_generator,
+                arrow_schema,
+                table_config,
+                primary_column_ids,
+            )
+            .await?,
+            inserted_rows: HashMap::new(),
+            row_converter,
+            primary_column_idx,
+        })
+    }
+
+    /// Write the batch.
+    pub async fn write(&mut self, batch: RecordBatch) -> Result<()> {
```

Comment:
The approach of separating `write` and `delete` maybe not quite convenient. Thinking about how risingwave will use it: the sink writer will accept a `StreamChunk`, it needs to call `write/delete` row by row. How about providing a method called `delte_write` like this:
```rust
pub async fn delta_write(&mut self, ops: IntArray, batch: RecordBatch) -> Result<()> {}
```

Where the ops array is for insert/delete operator, we can use 1 for insertion, 2 for deletion.